### PR TITLE
Highlight when a Parameter is controlled by a Send

### PR DESCRIPTION
**Pull request by qm210**
Hey,

for debugging larger patches, or just generally knowing what's going on, I found it useful to display parameters that are also targets of some "send" somewhere in a different color. tooltip over the value tells you even more information

image

PS: because I had this on a branch where I was generally improving (IMO) user experience (i.e. my own), this also has a minor fix that when you press the "Add Unit", the searchEditor is instantly active (i.e. you can start writing the unit name without clicking, I hate unnecessary clicks); and a quick styling fix that puts the operator comment into "quotes" for clarity.

---

on that occasion, (but these are basics that I could also find out by reading the source code) -- what actually happens when e.g. I have a gain that is 80 and I send a value to it with amount 128? will it then go to 128, or somehow, larger? and when zending a 0, will it stay at 80 or go to 0?

### CODE REVIEW

File: tracker/model.go [hunk 1]
```diff
@@ -590,3 +591,38 @@ func clamp(a, min, max int) int {
 	}
 	return a
 }
+
+func (m *Model) CollectSendsTo(param Parameter) iter.Seq[sointu.Unit] {
```
Comment (vsariola) [inline tracker/model.go:595]: This probably should not yield sointu.Unit as this will return the entire Parameters map for the GUI to fool around. Rather, returning a struct with whatever data you need for visualizing the send e.g. something like `struct { ID int; Amount int; SourceInstrName string ... }` would make sense here. This way we don't unnecessarily expose the Unit.Parameters map to the GUI for modification.

Comment (vsariola) [reply tracker/model.go:595]: Also, eventually, if we go with the cache approach, this function could be very cheap: just essentially pulling the values from the cache, which gets updated only when the patch changes.

File: tracker/model.go [hunk 2]
```diff
@@ -590,3 +591,38 @@ func clamp(a, min, max int) int {
...
+				}
+				if !yield(unit) {
+					return
+				}
+			}
+		}
+	}
+}
+
+func (m *Model) InstrumentForUnit(id int) *sointu.Instrument {
```
Comment (vsariola) [inline tracker/model.go:618]: You are returning *sointu.Instrument, but instr is in stack. So, when you return &instr, Go copies the value from stack to heap and returns a pointer to that, which means unnecessary heap allocation.

If you really would want to return a pointer to the instrument, &m.d.Song.Patch[...] would do the trick, but we really don't want to expose the m.d.Song directly to the GUI, because the whole point of Model package was to make sure gioui does not directly change the data without telling the Model.

Elsewhere, I even make a deep copy of the instrument to not expose the data to the gioui:

https://github.com/vsariola/sointu/blob/daf7fb1519856a0ab279815f19d4907bf48902e3/tracker/model.go#L407

Both places would benefit from figuring out how we efficiently tell the gioui side of things what sends point where, without exposing entire instruments to the gioui.

**Review by vsariola — COMMENTED** [review]
I could merge as it is, unless you immediately want to do the easy changes (returning value types instead of pointer type).

I can later refactor this for efficiency, so that some sort of cache is generated only when patch changes, and in that cache, we have stored which sends target what units and fast way (maps) to look this information up.

**Comment by vsariola** [comment]
> on that occasion, (but these are basics that I could also find out by reading the source code) -- what actually happens when e.g. I have a gain that is 80 and I send a value to it with amount 128? will it then go to 128, or somehow, larger? and when zending a 0, will it stay at 80 or go to 0?

Gain value of 80 is converted to float with x/128.0 so the gain is 0,625, and then if you send signal x to it with send amount 128, the resulting gain will be x+0.625 i.e. can very well go over 1. This has the implication that some units can create NaNs if a parameter is modulated and goes beyond the normal range.

**Comment by vsariola** [comment]
The force-pushed version from last night seems to include a stab at the cache, great! But it does not seem to work yet? At least I don't see the send targets highlighting anymore.

If this was a draft version and just saving evening's work, you might consider keeping these WIPs in another branch, because I get notifications when there is a new version in the branch associated with pull-requests and I am a bit unsure if that's an invitation to rereview or merge, or just that you were finished for the night.

That being said: I think it's a good idea to put all the cache / derived data (whatever we call it) in one place (e.g., derived.go)! I can eventually throw updatePatternUseCount there too, because it is conceptually similar cache.

**Comment by vsariola** [comment]
The only challenge with this approach is that we need to make sure that the caches are always regenerated correctly, but I don't see major bugs appearing even if these caches somehow manage to be out-of-date, as long as we use the caches just to display some helpful information in the GUI.

File: tracker/derived.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+)
+
+// public access functions
+
+func (m *Model) forUnitById(id int) *derivedForUnit {
+	forUnit, ok := m.derived.forUnit[id]
+	if !ok {
+		return nil
+	}
+	return &forUnit
```
Comment (vsariola) [inline tracker/derived.go:43]: This causes heap escaping because forUnit is a local variable and taking a pointer to a local variable requires moving the variable to heap before returning. One nuance of maps is that you cannot take address of a map value i.e. &m["key"] is an error. Thus, the return value of this probably should be derivedUnit, not *derivedUnit, and you might consider returning ok too in that case to indicate if the struct was populated.

File: tracker/derived.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+		return ""
+	}
+	return ft.title
+}
+
+// public getters with further model information
+
+func (m *Model) TracksWithSameInstrumentAsCurrent() []int {
+	currentTrack := m.d.Cursor.Track
+	return m.derived.forTrack[currentTrack].tracksWithSameInstrument
```
Comment (vsariola) [inline tracker/derived.go:90]: this will crash if m.derived.forTrack[currentTrack] manages to be nil pointer.

There is clear case when this can happen: when there is no tracks. The cursor track will stay at 0 even then.

File: tracker/derived.go [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+			tracksWithSameInstrument: slices.Collect(m.tracksWithSameInstrument(index)),
+			title:                    m.buildTrackTitle(index),
+		}
+	}
+}
+
+func (m *Model) updateDerivedPatchData() {
+	for _, instr := range m.d.Song.Patch {
+		for _, unit := range instr.Units {
+			m.derived.forUnit[unit.ID] = derivedForUnit{
```
Comment (vsariola) [inline tracker/derived.go:129]: While it's quite harmless that the maps are only generated once, it still can theoretically grow infinitely over time, if the unit IDs grow. Probably should just create a new map every update, or use [clear](https://tip.golang.org/ref/spec#Clear)

File: tracker/derived.go (hunk 4 above)
Comment (vsariola) [inline tracker/derived.go:90]: Forget what I said, this won't crash because it's a map of value types. So, this is fine. It'll just return nil slice if it cant find the key in the map.

File: tracker/derived.go (hunk 3 above)
Comment (vsariola) [reply tracker/derived.go:43]: To be clear: returning a small value type is usually faster than doing a heap allocation, and returning also ok or error to indicate if the struct is populated is idiomatic go

**Comment by qm210** [comment]
> > on that occasion, (but these are basics that I could also find out by reading the source code) -- what actually happens when e.g. I have a gain that is 80 and I send a value to it with amount 128? will it then go to 128, or somehow, larger? and when zending a 0, will it stay at 80 or go to 0?
> 
> Gain value of 80 is converted to float with x/128.0 so the gain is 0,625, and then if you send signal x to it with send amount 128, the resulting gain will be x+0.625 i.e. can very well go over 1. This has the implication that some units can create NaNs if a parameter is modulated and goes beyond the normal range.

Oh, that is interesting to know. Is that something were we would benefit if the UI is actually warning about such cases? (new issue?)

File: tracker/derived.go (hunk 4 above)
Comment (qm210) [reply tracker/derived.go:90]: ah. is there actually any difference between a map[int] and a normal slice, if all possible int keys are 0..<max value> anyway? what should we prefer?

Comment (vsariola) [reply tracker/derived.go:88]: Array loads and stores are a bit more performant, but this is insignificant. Maps are a bit more robust in the sense that if you try fetch a key that does not exist in the map, it returns the default value. Arrays can cause panics if the index is out of bounds. But checking that index is within bounds is also easy.

Tl;DR probably doesn't matter; given all indices are always linear range 0..N, arrays a bit more natural choice.

**Comment by vsariola** [comment]
> > Gain value of 80 is converted to float with x/128.0 so the gain is 0,625, and then if you send signal x to it with send amount 128, the resulting gain will be x+0.625 i.e. can very well go over 1. This has the implication that some units can create NaNs if a parameter is modulated and goes beyond the normal range.
> 
> Oh, that is interesting to know. Is that something were we would benefit if the UI is actually warning about such cases? (new issue?)

It's very hard to give accurate error messages because there is no guarantee that the signal sent is actually within -1..1 (for example: a simple LFO oscillator has the gain parameter to adjust it so it could be -0.5..0.5). Also, many units are just fine having their params go beyond 0..1, it's just some that aint.

Bottomline: we already warn when there's NaN; that's when we know things have gone really bad; everything before that it's "fine" and might be actually what the musician intended to do

File: tracker/model.go (hunk 1 above)
Comment (qm210) [reply tracker/model.go]: implemented that now

File: tracker/gioui/unit_editor.go [hunk 6]
```diff
@@ -240,24 +243,29 @@ type ParameterWidget struct {
...
 type ParameterStyle struct {
 	tracker *Tracker
 	w       *ParameterWidget
 	Theme   *material.Theme
 	Focus   bool
+	sends   []sointu.Unit
 }
 
 func (t *Tracker) ParamStyle(th *material.Theme, paramWidget *ParameterWidget) ParameterStyle {
+	sends := slices.Collect(t.Model.CollectSendsTo(paramWidget.Parameter))
```
[the remark this answers is not in the document]
Comment (qm210) [reply tracker/gioui/unit_editor.go]: see model.go l.237 and l.253, that should do it?

File: tracker/model.go (hunk 2 above)
Comment (qm210) [reply tracker/model.go]: that part is now unified in the new cache structure, no Copy() anymore

File: tracker/derived.go (hunk 3 above)
Comment (qm210) [reply tracker/derived.go]: should be implemented thoroughly now (in derived.go)

File: tracker/derived.go (hunk 5 above)
Comment (qm210) [reply tracker/derived.go:170]: clear() is done now

File: tracker/derived.go (hunk 4 above)
Comment (qm210) [reply tracker/derived.go:123]: forTrack / forPattern now use arrays

**Comment by vsariola** [comment]
Something is still off in the latest version:

image

Also, I managed to make the tracker crash:

```
...
C:/Users/sariola/go/pkg/mod/gioui.org@v0.7.1/op/clip/clip.go:105 +0x31
panic({0x7ff746319b40?, 0x7ff746287100?})
	C:/Program Files/Go/src/runtime/panic.go:785 +0x136
github.com/vsariola/sointu/tracker.(*Model).collectSendSources.func1(0xc003810ce0)
	c:/Users/sariola/git-repos/sointu/tracker/derived.go:214 +0x34e
slices.AppendSeq[...]({0x0, 0x0, 0x0}, 0xc002b24cb0)
	C:/Program Files/Go/src/slices/iter.go:50 +0x172
slices.Collect[...](0xc002b24cb0)
	C:/Program Files/Go/src/slices/iter.go:58 +0x47
...

```

The offending line: ```
				unitParam := sointu.FindParamForModulationPort(unit.Type, port)
				if unitParam.Name != paramName {
					continue
				}
```

There is no check for nil pointers. In general, returning a struct and ok bool is better, because the user then realizes that (s)he needs to consider what happens when ok is false.

So,
```
func FindParamForModulationPort(unitName string, index int) (up UnitParameter, ok bool)
```
would have avoided this crash (there's again a heap escape in the `			return &param` now, so would be better even)

**Comment by vsariola** [comment]
In FindParamForModulationPort, the param.CanModulate is handled incorrectly, because it returns the stereo parameter (which has CanModulate false) if the index is 0.

Correct:
```
func FindParamForModulationPort(unitName string, index int) (UnitParameter, bool) {
	unitType, ok := UnitTypes[unitName]
	if !ok {
		return UnitParameter{}, false
	}
	for _, param := range unitType {
		if !param.CanModulate {
			continue
		}
		if index == 0 {
			return param, true
		}
		index--
	}
	return UnitParameter{}, false
}
```

**Comment by vsariola** [comment]
You can see that when selecting envelope attack as the send target, it does not highlight:
image
image

**Comment by vsariola** [comment]
LGTM, thanks! Merged, thank you for your contribution ❤️
